extension/clientcredentialsauthextension/extension.go

```diff
@@ -0,0 +1,110 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clientcredentialsauthextension
+
+import (
+	"context"
+	"net/http"
+
+	"go.uber.org/zap"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+	"google.golang.org/grpc/credentials"
+	grpcOAuth "google.golang.org/grpc/credentials/oauth"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configauth"
+)
+
+// ClientCredentialsAuthenticator provides implementation for providing client authentication using OAuth2 client credentials
+// workflow for both gRPC and HTTP clients.
+type ClientCredentialsAuthenticator struct {
+	clientCredentials *clientcredentials.Config
+	logger            *zap.Logger
+	client            *http.Client
+}
+
+// ClientCredentialsAuthenticator implements both HTTPClientAuth and GRPCClientAuth
+var (
+	_ configauth.HTTPClientAuthenticator = (*ClientCredentialsAuthenticator)(nil)
+	_ configauth.GRPCClientAuthenticator = (*ClientCredentialsAuthenticator)(nil)
+)
+
+func newClientCredentialsExtension(cfg *Config, logger *zap.Logger) (*ClientCredentialsAuthenticator, error) {
+	if cfg.ClientID == "" {
+		return nil, errNoClientIDProvided
+	}
+	if cfg.ClientSecret == "" {
+		return nil, errNoClientSecretProvided
+	}
+	if cfg.TokenURL == "" {
+		return nil, errNoTokenURLProvided
+	}
+
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+
+	if cfg.TLSSetting != nil {
+		tlsCfg, err := cfg.TLSSetting.LoadTLSConfig()
+		if err != nil {
+			return nil, err
+		}
+		if tlsCfg != nil {
+			transport.TLSClientConfig = tlsCfg
+		}
+	}
+
+	return &ClientCredentialsAuthenticator{
+		clientCredentials: &clientcredentials.Config{
+			ClientID:     cfg.ClientID,
+			ClientSecret: cfg.ClientSecret,
+			TokenURL:     cfg.TokenURL,
+			Scopes:       cfg.Scopes,
+		},
+		logger: logger,
+		client: &http.Client{
+			Transport: transport,
+			Timeout:   cfg.Timeout,
+		},
+	}, nil
+}
+
+// Start for ClientCredentialsAuthenticator extension does nothing
+func (o *ClientCredentialsAuthenticator) Start(_ context.Context, _ component.Host) error {
+	return nil
+}
+
+// Shutdown for ClientCredentialsAuthenticator extension does nothing
+func (o *ClientCredentialsAuthenticator) Shutdown(_ context.Context) error {
+	return nil
+}
+
+// RoundTripper returns oauth2.Transport, an http.RoundTripper that performs "client-credential" OAuth flow and
+// also auto refreshes OAuth tokens as needed.
+func (o *ClientCredentialsAuthenticator) RoundTripper(base http.RoundTripper) (http.RoundTripper, error) {
+	ctx := context.WithValue(context.Background(), oauth2.HTTPClient, o.client)
```

Comment:
Apologies if I asked this before during a previous review, but is there a timeout being set somewhere in this call chain?

Reply:
@jpkrohling  we are passing the timeout as a configuration and we use it to build `o.client ` with it. In the case where user ignores to set the timeout (defaults to 0) - the client doesn't have a timeout. This is inline with what we do in http client in confighttp.